dataRetriever/dataPool/proofsCache/proofsCache.go

```diff
@@ -0,0 +1,88 @@
+package proofscache
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/data"
+)
+
+type proofNonceMapping struct {
+	headerHash string
+	nonce      uint64
+}
+
+type proofsCache struct {
+	mutProofsCache sync.RWMutex
+	proofsByNonce  []*proofNonceMapping
+	proofsByHash   map[string]data.HeaderProofHandler
+}
+
+func newProofsCache() *proofsCache {
+	return &proofsCache{
+		mutProofsCache: sync.RWMutex{},
+		proofsByNonce:  make([]*proofNonceMapping, 0),
+		proofsByHash:   make(map[string]data.HeaderProofHandler),
+	}
+}
+
+func (pc *proofsCache) getProofByHash(headerHash []byte) (data.HeaderProofHandler, error) {
+	pc.mutProofsCache.RLock()
+	defer pc.mutProofsCache.RUnlock()
+
+	proof, ok := pc.proofsByHash[string(headerHash)]
+	if !ok {
+		return nil, ErrMissingProof
+	}
+
+	return proof, nil
+}
+
+func (pc *proofsCache) getAllProofs() map[string]data.HeaderProofHandler {
+	pc.mutProofsCache.RLock()
+	defer pc.mutProofsCache.RUnlock()
+
+	return pc.proofsByHash
+}
+
+func (pc *proofsCache) addProof(proof data.HeaderProofHandler) {
+	if proof == nil {
+		return
+	}
+
+	pc.mutProofsCache.Lock()
+	defer pc.mutProofsCache.Unlock()
+
+	pc.proofsByNonce = append(pc.proofsByNonce, &proofNonceMapping{
```

Comment:
maybe check if proof already exists for this proof.GetHeaderHash.. we don't want to overwrite the first proof saved

Reply:
done